src/rust/engine/process_execution/src/docker.rs

```diff
@@ -435,58 +461,123 @@ impl<'a> super::CommandRunner for CommandRunner<'a> {
 
 #[async_trait]
 impl<'a> CapturedWorkdir for CommandRunner<'a> {
-  type WorkdirToken = String;
+  type WorkdirToken = (String, String);
 
   async fn run_in_workdir<'s, 'c, 'w, 'r>(
     &'s self,
-    context: &'c Context,
+    _context: &'c Context,
     _workdir_path: &'w Path,
-    sandbox_path_in_container: Self::WorkdirToken,
+    (container_id, working_dir): Self::WorkdirToken,
     req: Process,
     _exclusive_spawn: bool,
   ) -> Result<BoxStream<'r, Result<ChildOutput, String>>, String> {
     let docker = self.docker.get().await?;
 
-    let env = req
-      .env
-      .iter()
-      .map(|(key, value)| format!("{key}={value}"))
-      .collect::<Vec<_>>();
+    Command::new(req.argv)
+      .env(req.env)
+      .working_dir(working_dir)
+      .spawn(docker, container_id)
+      .await
+  }
 
-    let working_dir = req
-      .working_directory
-      .map(|relpath| Path::new(&sandbox_path_in_container).join(relpath))
-      .unwrap_or_else(|| Path::new(&sandbox_path_in_container).to_path_buf())
-      .into_os_string()
-      .into_string()
-      .map_err(|s| {
-        format!("Unable to convert working directory due to non UTF-8 characters: {s:?}")
-      })?;
+  async fn prepare_workdir_for_capture(
+    &self,
+    _context: &Context,
+    _workdir_path: &Path,
+    (container_id, working_dir): Self::WorkdirToken,
+    req: &Process,
+  ) -> Result<(), String> {
+    // Docker on Linux will frequently produce root-owned output files in bind mounts, because we
```

Comment:
Just reading the comment as it is I think you can strike Linux from the sentence. This happens on Mac too right? The missing information is how do bind mounts work on Mac. Do they magically do this perm fixing already in their journey to get files from the container transferred to the host?

Reply:
It doesn't happen on macOS: I confirmed that files are created as the user running Docker.

Reply:
Can you maybe note this in the comment. Stepping back - my rough impression is we generally don't know how Docker works. We're paying for that and patching as we learn more. The issue is not knowing what we're doing (I'm used to this over in Pex / PyPA land!). It would seem excellent to be very diligent about calling shots. We know X by experiment, but not why or if guaranteed, etc.